iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift

```diff
@@ -0,0 +1,304 @@
+//
+//  TimelineLayout.swift
+//  TimelineExperiment
+//
+//  Created by Martin Young on 11/16/21.
+//
+
+import Foundation
+import UIKit
+
+protocol TimelineCollectionViewLayoutDataSource: AnyObject {
+    func getMessage(at indexPath: IndexPath) -> Messageable?
+}
+
+/// A custom layout for conversation messages. Up to two message cell sections are each displayed as a stack along the z axis.
+/// The stacks appear similar to Apple's Time Machine interface, with the newest message in front and older messages going out into the distance.
+/// As the collection view scrolls up and down, the messages move away or toward the user.
+class TimelineCollectionViewLayout: UICollectionViewLayout {
+
+    private typealias SectionIndex = Int
+
+    weak var dataSource: TimelineCollectionViewLayoutDataSource?
+
+    /// The size of the cells.
+    var itemSize = CGSize(width: 200, height: 100)
+
+    /// A cache of item layout attributes so they don't have to be recalculated.
+    private var cellLayoutAttributes: [IndexPath : UICollectionViewLayoutAttributes] = [:]
+    /// A dictionary of z ranges for all the items. A z range represents the range that each item will be frontmost in its section
+    /// and its scale will be unaltered.
+    private var zRangesDict: [IndexPath : Range<CGFloat>] = [:]
+    /// The current position along the Z axis. This is based off of the collectionview's Y content offset.
+    private var zPosition: CGFloat {
+        return self.collectionView?.contentOffset.y ?? 0
+    }
+
+    private var sectionCount: Int {
+        return self.collectionView?.numberOfSections ?? 0
+    }
+    private func numberOfItems(inSection section: Int) -> Int {
+        guard section < self.sectionCount else { return 0 }
+        return self.collectionView?.numberOfItems(inSection: section) ?? 0
+    }
+
+    // MARK: - UICollectionViewLayout Overrides
+
+    override var collectionViewContentSize: CGSize {
+        get {
+            guard let collectionView = collectionView, collectionView.frame != .zero else { return .zero }
+
+            let itemCount = CGFloat(self.numberOfItems(inSection: 0) + self.numberOfItems(inSection: 1))
+            var height = (itemCount - 1) * self.itemSize.height
+            height += collectionView.bounds.height
+            return CGSize(width: collectionView.bounds.width, height: height)
```

Comment:
Make the content offset big enough to scroll from the beginning end of the items.